src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -407,29 +410,33 @@ void Chttp2ServerListener::ActiveConnection::HandshakingState::Start(
                              OnHandshakeDone, this);
 }
 
-void Chttp2ServerListener::ActiveConnection::HandshakingState::OnTimeout(
-    void* arg, grpc_error_handle error) {
-  HandshakingState* self = static_cast<HandshakingState*>(arg);
-  // Note that we may be called with absl::OkStatus() when the timer fires
-  // or with an error indicating that the timer system is being shut down.
-  if (error != absl::CancelledError()) {
+void Chttp2ServerListener::ActiveConnection::HandshakingState::OnTimeout() {
+  grpc_chttp2_transport* transport = nullptr;
+  {
+    MutexLock lock(&connection_->mu_);
+    if (timer_handle_.has_value()) {
+      transport = connection_->transport_;
+      timer_handle_.reset();
+    }
+  }
+  if (transport) {
     grpc_transport_op* op = grpc_make_transport_op(nullptr);
     op->disconnect_with_error = GRPC_ERROR_CREATE(
         "Did not receive HTTP/2 settings before handshake timeout");
-    grpc_chttp2_transport* transport = nullptr;
-    {
-      MutexLock lock(&self->connection_->mu_);
-      transport = self->connection_->transport_;
-    }
     grpc_transport_perform_op(&transport->base, op);
   }
-  self->Unref();
 }
 
 void Chttp2ServerListener::ActiveConnection::HandshakingState::
     OnReceiveSettings(void* arg, grpc_error_handle /* error */) {
   HandshakingState* self = static_cast<HandshakingState*>(arg);
-  grpc_timer_cancel(&self->timer_);
```

Comment:
Check my logic, please, on behavior deviations: Even with `grpc_timer_cancel` called here, I think there was possibly a race in the previous implementation where `OnTimeout` may have been called with an OK status already (imagine ~concurrently with `OnReceiveSettings`), in which case the `grpc_transport_perform_op` in OnTimeout would have proceeded. It seems like this was previously incorrect behavior, and this PR fixes that by resetting the timer handle under a mutex here, and checking it in `OnTimeout` before committing to performing the op.

Reply:
Good case! I didn't think about this case. I think it's possible. But I'm wondering what the correct behavior in this case would be. I feel like from the server's point of view it might rightfully think that the client has timed out to send the settings, send goaway and disconnect while at the same time the client sends the settings...

Reply:
In this race, presuming timeout and receipt of settings are happening concurrently on the server, I think I would err on the side of leniency towards the client, which is what this PR is doing. But I'm not 100% certain that's correct either.

@markdroth would you mind sharing your insights?

Reply:
Yes, this looks right to me.  If we've actually seen the settings, then we shouldn't close the connection when the callback runs.

Looking at the existing code, I'm actually surprised this built, because `timer_` is marked as being guarded by the mutex, but here we were cancelling the timer without holding the mutex.  Not sure why that was working, but I guess it doesn't matter now.

Reply:
There is always a race given that the network is involved and that the SETTINGS frame can be received at any time, but I think this is a good change. Speaking purely from the Chttp2Server's perspective, if it has already seen the settings, it should ignore the timer even if the timer was fired with an OK status.